net-tools/netwatch/src/udp.rs

```diff
@@ -100,48 +725,152 @@ impl UdpSocket {
                 local_addr.port(),
             );
         }
-        Ok(UdpSocket(Some(socket)))
-    }
-}
 
-#[cfg(unix)]
-impl std::os::fd::AsFd for UdpSocket {
-    fn as_fd(&self) -> std::os::fd::BorrowedFd<'_> {
-        self.0.as_ref().expect("not dropped").as_fd()
+        Ok(Self::Connected {
+            socket,
+            state: socket_state,
+            addr: local_addr,
+        })
     }
-}
 
-#[cfg(windows)]
-impl std::os::windows::io::AsSocket for UdpSocket {
-    fn as_socket(&self) -> std::os::windows::io::BorrowedSocket<'_> {
-        self.0.as_ref().expect("not dropped").as_socket()
-    }
-}
+    fn rebind(&mut self) -> Result<()> {
+        let addr = match self {
+            Self::Connected { addr, .. } => *addr,
+            Self::Closed => {
+                bail!("socket is closed and cannot be rebound");
+            }
+        };
+        debug!("rebinding {}", addr);
 
-impl From<tokio::net::UdpSocket> for UdpSocket {
-    fn from(socket: tokio::net::UdpSocket) -> Self {
-        Self(Some(socket))
+        *self = SocketState::Closed;
+        *self = Self::bind(addr)?;
+
+        Ok(())
     }
-}
 
-impl std::ops::Deref for UdpSocket {
-    type Target = tokio::net::UdpSocket;
+    fn is_closed(&self) -> bool {
+        matches!(self, Self::Closed)
+    }
 
-    fn deref(&self) -> &Self::Target {
-        self.0.as_ref().expect("only removed on drop")
+    fn close(&mut self) -> Option<(tokio::net::UdpSocket, quinn_udp::UdpSocketState)> {
+        match std::mem::replace(self, SocketState::Closed) {
+            Self::Connected { socket, state, .. } => Some((socket, state)),
+            Self::Closed => None,
+        }
     }
 }
 
 impl Drop for UdpSocket {
     fn drop(&mut self) {
-        let std_sock = self.0.take().expect("not yet dropped").into_std();
+        trace!("dropping UdpSocket");
+        match self.socket.write().unwrap().close() {
+            Some((socket, _)) => {
+                if let Ok(handle) = tokio::runtime::Handle::try_current() {
+                    // No wakeup after dropping write lock here, since we're getting dropped.
+                    // this will be empty if `close` was called before
+                    let std_sock = socket.into_std();
+                    handle.spawn_blocking(move || {
+                        // Calls libc::close, which can block
+                        drop(std_sock);
+                    });
+                }
+            }
+            None => {}
+        }
```

Comment:
I agree with clippy here. This can be an `if-let`:
```suggestion
        if let Some((socket, _)) = self.socket.write().unwrap().close() {
            if let Ok(handle) = tokio::runtime::Handle::try_current() {
                // No wakeup after dropping write lock here, since we're getting dropped.
                // this will be empty if `close` was called before
                let std_sock = socket.into_std();
                handle.spawn_blocking(move || {
                    // Calls libc::close, which can block
                    drop(std_sock);
                });
            }
        }
```

Reply:
I hate if let.. it always makes me feel like I forgot sth